fapolicy_analyzer/ui/fapd_manager.py

```diff
@@ -0,0 +1,215 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import logging
+import os
+import subprocess
+
+
+from datetime import datetime as DT
+from enum import Enum
+from fapolicy_analyzer import Handle
+from threading import Lock
+from time import time, sleep
+
+
+class ServiceStatus(Enum):
+    FALSE = False
+    TRUE = True
+    UNKNOWN = None
+
+
+class FapdMode(Enum):
+    DISABLED = 0
+    ONLINE = 1
+    PROFILING = 2
+
+
+class FapdManager():
+    def __init__(self, bSuFapdControl=True):
+        logging.debug("FapdManager::__init__(self)")
+        self._fapd_status = ServiceStatus.UNKNOWN
+        self._fapd_monitoring = False
+        self._fapd_ref = Handle("fapolicyd")
+        self._fapd_profiler_pid = None
+        self._fapd_control_enabled = bSuFapdControl
+        self._fapd_lock = Lock()
+        self.mode = FapdMode.DISABLED
+        self.procProfile = None
+        self._fapd_profiling_timestamp = None
+
+        # Verify that fapolicyd is intalled, if so initialize fapd status
+        self.initial_daemon_status()
+
+        # ToDo: Verify that FAPD_LOGPATH works in the pkexec environment
+        if os.environ.get("FAPD_LOGPATH"):
+            self.fapd_profiling_stdout = os.environ.get("FAPD_LOGPATH") + ".stdout"
+            self.fapd_profiling_stderr = os.environ.get("FAPD_LOGPATH") + ".stderr"
+        else:
+            self.fapd_profiling_stdout = None
+            self.fapd_profiling_stderr = None
+
+    # ############### Public Interface ###############
+    def start_online_fapd(self):
+        # ToDo: Add logic to that online and profiling instances are mutex
+        self.set_mode(FapdMode.ONLINE)
+        self._start()
+
+    def stop_online_fapd(self):
+        # ToDo: Add logic to that online and profiling instances are mutex
+        self.set_mode(FapdMode.ONLINE)
+        self._stop()
+
+    def status_online_fapd(self):
+        # ToDo: Add logic to that online and profiling instances are mutex
+        # ToDo: Consider returning a tuple with status and instance info
+        self.set_mode(FapdMode.ONLINE)
+        return self._status()
+
+    def start_profiling_fapd(self):
+        # ToDo: Add logic to that online and profiling instances are mutex
+        self.set_mode(FapdMode.PROFILING)
+        self._start()
+
+    def stop_profiling_fapd(self):
+        # ToDo: Add logic to that online and profiling instances are mutex
+        self.set_mode(FapdMode.PROFILING)
+        self._stop()
+
+    def status_profiling_fapd(self):
+        # ToDo: Add logic to that online and profiling instances are mutex
+        # ToDo: Consider returning a tuple with status and instance info
+        self.set_mode(FapdMode.PROFILING)
```

Comment:
Should requesting the status change the `mode` of the manager?

Reply:
TBD. The original intention of the mode argument was to communicate with either fapolicyd instance with the same interface, so technically it would be possible for the client programmer to sequentially query the two fapolicyd instances. e.g. fapd status, and update the GUI appropriately. But... I'm not sure if that's the way to go. Will code up the control dlg and check the workflow.

Reply:
Left the self.mode attribute as part of the public interface, for the client code to select the desired target fapd instance to direct start/stop/status requests. Added two new attributes to represent the current active fapd instance and the prior instance (if any). The intention is to return the online fapd instance to the same pre-profiling run state after profiling is completed.